### PR TITLE
PPLUS-1591: Add documentation link to CLI report output

### PR DESCRIPTION
**Pull request by DmytroKlymanSpryker**
Provide documentation link for evaluator output in verbose mode (if `-v` option is used).

### CODE REVIEW

File: src/CodeCompliance/Domain/Checks/NotUnique/ModuleName.php [hunk 1]
```diff
@@ -46,6 +47,14 @@ public function getGuideline(): string
         return 'Module %s has to have project prefix, like %s%s.';
     }
 
+    /**
+     * @return string
+     */
+    public function getDocumentationUrl(): string
+    {
+        return static::DOCUMENTATION_BASE_URL . 'entity-name-is-not-unique.html#module-name-is-not-unique';
```
Comment (kalaputsik) [inline src/CodeCompliance/Domain/Checks/NotUnique/ModuleName.php:55]: looks like correct link
https://docs.spryker.com/docs/scos/dev/guidelines/keeping-a-project-upgradable/upgradability-guidelines/entity-name-is-not-unique.html#example-of-related-error-in-the-evaluator-output-module-name-is-not-unique

File: src/CodeCompliance/Domain/Checks/PrivateApi/Extended/Extended.php [hunk 2]
```diff
@@ -35,6 +36,14 @@ public function getGuideline(): string
         return 'Please avoid extension of the PrivateApi %s in %s';
     }
 
+    /**
+     * @return string
+     */
+    public function getDocumentationUrl(): string
+    {
+        return static::DOCUMENTATION_BASE_URL . 'private-api-method-is-overridden-on-the-project-level.html';
```
Comment (kalaputsik) [inline src/CodeCompliance/Domain/Checks/PrivateApi/Extended/Extended.php:44]: looks like the correct link
https://docs.spryker.com/docs/scos/dev/guidelines/keeping-a-project-upgradable/upgradability-guidelines/private-api-is-extended.html#example-of-code-that-causes-an-upgradability-error-extending-a-private-api-form-class

File: src/CodeCompliance/Domain/Checks/PrivateApi/Used/DependencyInBusinessModel.php [hunk 3]
```diff
@@ -36,6 +37,14 @@ public function getGuideline(): string
         return 'Please avoid usage of %s in %s';
     }
 
+    /**
+     * @return string
+     */
+    public function getDocumentationUrl(): string
+    {
+        return static::DOCUMENTATION_BASE_URL . 'private-api-is-extended.html';
```
Comment (kalaputsik) [inline src/CodeCompliance/Domain/Checks/PrivateApi/Used/DependencyInBusinessModel.php:45]: might be more appropriate
https://docs.spryker.com/docs/scos/dev/guidelines/keeping-a-project-upgradable/upgradability-guidelines/private-api-is-extended.html#example-of-code-that-causes-an-upgradability-error-extending-a-private-api-business-model

File: src/CodeCompliance/Domain/Checks/PrivateApi/Used/DependencyProvider.php [hunk 4]
```diff
@@ -32,6 +33,14 @@ public function getGuideline(): string
         return 'Please avoid usage of %s::%s in %s';
     }
 
+    /**
+     * @return string
+     */
+    public function getDocumentationUrl(): string
+    {
+        return static::DOCUMENTATION_BASE_URL . 'private-api-is-extended.html';
```
Comment (kalaputsik) [inline src/CodeCompliance/Domain/Checks/PrivateApi/Used/DependencyProvider.php:41]: https://docs.spryker.com/docs/scos/dev/guidelines/keeping-a-project-upgradable/upgradability-guidelines/private-api-is-extended.html#example-of-code-that-causes-an-upgradability-error-extending-a-private-api-dependency-provider

File: src/CodeCompliance/Domain/Checks/PrivateApi/Used/Facade.php [hunk 5]
```diff
@@ -41,6 +42,14 @@ public function getGuideline(): string
         return 'Please avoid usage of %s(...) in %s';
     }
 
+    /**
+     * @return string
+     */
+    public function getDocumentationUrl(): string
+    {
+        return static::DOCUMENTATION_BASE_URL . 'private-api-is-extended.html';
```
Comment (kalaputsik) [inline src/CodeCompliance/Domain/Checks/PrivateApi/Used/Facade.php:50]: not sure, but might be better
https://docs.spryker.com/docs/scos/dev/guidelines/keeping-a-project-upgradable/upgradability-guidelines/private-api-is-extended.html#example-of-code-that-causes-an-upgradability-error-extending-a-private-api-dependency-provider

File: src/CodeCompliance/Domain/Checks/PrivateApi/Used/ObjectIsInitializedInBusinessModel.php [hunk 6]
```diff
@@ -33,6 +34,14 @@ public function getGuideline(): string
         return 'Please avoid usage of %s in %s';
     }
 
+    /**
+     * @return string
+     */
+    public function getDocumentationUrl(): string
+    {
+        return static::DOCUMENTATION_BASE_URL . 'private-api-is-extended.html';
```
Comment (kalaputsik) [inline src/CodeCompliance/Domain/Checks/PrivateApi/Used/ObjectIsInitializedInBusinessModel.php:42]: what about this?
https://docs.spryker.com/docs/scos/dev/guidelines/keeping-a-project-upgradable/upgradability-guidelines/private-api-is-used-on-the-project-level.html#example-of-code-that-causes-an-upgradability-error

File: src/CodeCompliance/Domain/Checks/NotUnique/ModuleName.php (hunk 1 above)
Comment (DmytroKlymanSpryker) [reply src/CodeCompliance/Domain/Checks/NotUnique/ModuleName.php:55]: In this case - we have a mistake in documentation at the moment. And the section for this error description has the wrong name. I fixed it inside of documentation - https://github.com/spryker/spryker-docs/pull/1262 . PR is waiting for approval.

So I put the address here that will be correct in the future :) At the moment the link for that is: 
https://docs.spryker.com/docs/scos/dev/guidelines/keeping-a-project-upgradable/upgradability-guidelines/entity-name-is-not-unique.html#constant-name-is-not-unique-1

So it contains the description and right after description - examples.

File: src/CodeCompliance/Domain/Checks/PrivateApi/Used/ObjectIsInitializedInBusinessModel.php (hunk 6 above)
Comment (DmytroKlymanSpryker) [reply src/CodeCompliance/Domain/Checks/PrivateApi/Used/ObjectIsInitializedInBusinessModel.php:42]: +

File: src/CodeCompliance/Domain/Checks/PrivateApi/Used/Facade.php (hunk 5 above)
Comment (DmytroKlymanSpryker) [reply src/CodeCompliance/Domain/Checks/PrivateApi/Used/Facade.php:50]: In this case it looks like we need more specific example for this rule check. I create another ticket for that - https://spryker.atlassian.net/browse/PPLUS-1626

And till that I leave a general link in this case if it's fine for you

File: src/CodeCompliance/Domain/Checks/PrivateApi/Used/DependencyInBusinessModel.php (hunk 3 above)
Comment (DmytroKlymanSpryker) [reply src/CodeCompliance/Domain/Checks/PrivateApi/Used/DependencyInBusinessModel.php:45]: \+

File: src/CodeCompliance/Domain/Checks/PrivateApi/Extended/Extended.php (hunk 2 above)
Comment (DmytroKlymanSpryker) [reply src/CodeCompliance/Domain/Checks/PrivateApi/Extended/Extended.php:44]: I am not sure in this case. The link for `PrivateApi:Extension` on this page: https://docs.spryker.com/docs/scos/dev/guidelines/keeping-a-project-upgradable/upgradability-guidelines/upgradability-guidelines.html

goes on the page that I use in code. If you think, that it makes more sense to use another page, I think we need to change it also in current documentation. I can put it also to the new ticket https://spryker.atlassian.net/browse/PPLUS-1626

File: src/Upgrader/Report/Service/ReportService.php [hunk 7]
```diff
@@ -71,8 +83,15 @@ public function report(): ?array
         $messageSeparator = str_pad('', 100, '-');
         foreach ($output[static::KEY_VIOLATIONS] as $violations) {
```
Comment (VladislavStrelchenko) [inline src/Upgrader/Report/Service/ReportService.php:84]: Please change `as $violations` to `as $violation`

File: src/Upgrader/Report/Service/ReportService.php [hunk 8]
```diff
@@ -71,8 +83,15 @@ public function report(): ?array
...
             $key = $violations[static::KEY_PRODUCED_BY];
-            $keySeparatorLength = str_pad('', strlen($key), '-');
-            $messages[] = $key . ' ' . $violations[static::KEY_MESSAGE] . PHP_EOL . $keySeparatorLength . ' ' . $messageSeparator;
+            $message = $key . ' ' . $violations[static::KEY_MESSAGE] . PHP_EOL;
```
Comment (VladislavStrelchenko) [inline src/Upgrader/Report/Service/ReportService.php:86]: From my perspective, better to add `getMessage` method 

`    public function report(bool $isVerbose = false): ?array
    {
        $path = getcwd() . sprintf(static::FILE_PATH_PATTERN, AnalyzeTask::ID_ANALYZE_TASK);

        if (!file_exists($path)) {
            return null;
        }

        $output = Yaml::parseFile($path);

        $messages = [];
        $messageSeparator = str_pad('', 100, '-');
        foreach ($output[static::KEY_VIOLATIONS] as $violation) {
            $messages[] = sprintf('%s %s', $this->getMessage($violation, $isVerbose), $messageSeparator);
        }

        return $messages;
    }

    /**
     * @param array $violations
     * @param bool $isVerbose
     * @return string|null
     */
    protected function getMessage(array $violations, bool $isVerbose = false): ?string
    {
        $key = $violations[static::KEY_PRODUCED_BY];
        $message = $key . ' ' . $violations[static::KEY_MESSAGE] . PHP_EOL;

        if ($isVerbose) {
            $docUrl = $violations[static::KEY_ADDITIONAL_ATTRIBUTES][static::KEY_ATTRIBUTE_DOCUMENTATION] ?? '';
            $message .= str_pad('', strlen($key) + 1, ' ') . '💡More information: ' . $docUrl . PHP_EOL;
        }

        $message .= str_pad('', strlen($key), '-');
        
        return $message;
    }`

File: src/CodeCompliance/Domain/Checks/PrivateApi/Used/PersistenceInBusinessModel.php [hunk 9]
```diff
@@ -90,7 +99,9 @@ public function getViolations(): array
                         )
                     ) {
                         $guideline = sprintf($this->getGuideline(), $methodReflection->getDeclaringClass()->getName(), $methodName, $source->getClassName());
-                        $violations[] = new Violation((string)(new Id()), $guideline, $this->getName());
+                        $violations[] = new Violation((string)(new Id()), $guideline, $this->getName(), ViolationInterface::SEVERITY_ERROR, [
+                            'documentation' => $this->getDocumentationUrl(),
```
Comment (VladislavStrelchenko) [inline src/CodeCompliance/Domain/Checks/PrivateApi/Used/PersistenceInBusinessModel.php:103]: Better to create const for `documentation` key on CodeCompliance level

File: src/CodeCompliance/Domain/AbstractCodeComplianceCheck.php [hunk 10]
```diff
@@ -15,6 +15,11 @@
 
 abstract class AbstractCodeComplianceCheck implements CodeComplianceCheckInterface
 {
+    /**
+     * @var string
+     */
+    protected const DOCUMENTATION_BASE_URL = 'https://docs.spryker.com/docs/scos/dev/guidelines/keeping-a-project-upgradable/upgradability-guidelines/';
```
Comment (VladislavStrelchenko) [inline src/CodeCompliance/Domain/AbstractCodeComplianceCheck.php:21]: What do you think about moving it to some configuration file
such as  Upgrader\Configuration\ConfigurationProvider or a new configuration file on CodeCompliance level

File: src/Upgrader/Report/Service/ReportService.php [hunk 11]
```diff
@@ -71,8 +83,15 @@ public function report(): ?array
...
+
+            if ($isVerbose) {
+                $docUrl = $violations[static::KEY_ADDITIONAL_ATTRIBUTES][static::KEY_ATTRIBUTE_DOCUMENTATION] ?? '';
+                $message .= str_pad('', strlen($key) + 1, ' ') . '💡More information: ' . $docUrl . PHP_EOL;
```
[the remark this answers is not in the document]
Comment (DmytroKlymanSpryker) [reply src/Upgrader/Report/Service/ReportService.php]: \+

File: src/CodeCompliance/Domain/Checks/PrivateApi/Used/PersistenceInBusinessModel.php (hunk 9 above)
Comment (DmytroKlymanSpryker) [reply src/CodeCompliance/Domain/Checks/PrivateApi/Used/PersistenceInBusinessModel.php]: \+

File: src/CodeCompliance/Domain/AbstractCodeComplianceCheck.php (hunk 10 above)
Comment (DmytroKlymanSpryker) [reply src/CodeCompliance/Domain/AbstractCodeComplianceCheck.php:21]: Yes, I wanted to store it somewhere in the configuration, but didn't find a place that is on my opinion suitable for that. But I saw before only infrastructure config providers. `Upgrader\Configuration\ConfigurationProvider` looks more suitable for that, I'll have a look

**Review by VladislavStrelchenko — APPROVED** [review]

**Review by kalaputsik — APPROVED** [review]

**Pull request merged**